lib/events.js

```diff
@@ -315,16 +315,17 @@ EventEmitter.prototype.removeListener =
       if (events === undefined)
         return this;
 
-      list = events[type];
+      list = events.get(type);
       if (list === undefined)
         return this;
 
       if (list === listener || list.listener === listener) {
-        if (--this._eventsCount === 0)
-          this._events = Object.create(null);
-        else {
-          delete events[type];
-          if (events.removeListener)
+        if (--this._eventsCount === 0) {
+          // this._events.clear();
+          this._events = new Map();
```

Comment:
What happens if we remove the `if` and always `delete` ?

Reply:
I've tried that initially but there is a significant regression for `ee-once` if we do that, seems that either Map creation is cheaper or there is some missed optimization. Maybe someone can take a look.

```
➔ dev/node/node cat compare-events-pr-1.csv | Rscript benchmark/compare.R
                                                     confidence improvement accuracy (*)    (**)   (***)
 events/ee-add-remove.js n=1000000                                   3.38 %       ±4.91%  ±6.73%  ±9.16%
 events/ee-emit.js listeners=1 argc=0 n=2000000                     -0.73 %      ±13.90% ±19.27% ±26.76%
 events/ee-emit.js listeners=1 argc=10 n=2000000            ***     15.68 %       ±7.91% ±10.91% ±15.04%
 events/ee-emit.js listeners=1 argc=2 n=2000000                      8.14 %       ±9.82% ±13.45% ±18.34%
 events/ee-emit.js listeners=1 argc=4 n=2000000                      9.24 %       ±9.50% ±13.11% ±18.05%
 events/ee-emit.js listeners=10 argc=0 n=2000000                    -0.54 %      ±10.02% ±13.75% ±18.80%
 events/ee-emit.js listeners=10 argc=10 n=2000000                    1.89 %       ±8.26% ±11.33% ±15.48%
 events/ee-emit.js listeners=10 argc=2 n=2000000                    -2.79 %      ±11.40% ±15.66% ±21.44%
 events/ee-emit.js listeners=10 argc=4 n=2000000                     2.88 %       ±6.85%  ±9.38% ±12.78%
 events/ee-emit.js listeners=5 argc=0 n=2000000                      0.59 %      ±11.30% ±15.49% ±21.11%
 events/ee-emit.js listeners=5 argc=10 n=2000000                     2.33 %       ±9.27% ±12.72% ±17.36%
 events/ee-emit.js listeners=5 argc=2 n=2000000                      0.10 %      ±11.37% ±15.61% ±21.33%
 events/ee-emit.js listeners=5 argc=4 n=2000000                     -2.23 %       ±9.85% ±13.50% ±18.41%
 events/ee-listener-count-on-prototype.js n=50000000        ***     31.36 %       ±8.61% ±12.06% ±17.04%
 events/ee-listeners-many.js n=5000000                               0.69 %      ±12.55% ±17.44% ±24.30%
 events/ee-listeners.js n=5000000                                   -2.08 %       ±8.51% ±11.78% ±16.31%
 events/ee-once.js n=20000000                               ***    -26.48 %       ±5.40%  ±7.42% ±10.16%
```

diff:
```patch

@ events.js:323 @ EventEmitter.prototype.removeListener =
         return this;

       if (list === listener || list.listener === listener) {
-        if (--this._eventsCount === 0) {
-          // this._events.clear();
-          this._events = new Map();
-        } else {
-          events.delete(type);
-          if (events.has('removeListener'))
-            this.emit('removeListener', type, list.listener || listener);
-        }
+        --this._eventsCount;
+        events.delete(type);
+        if (events.has('removeListener'))
+          this.emit('removeListener', type, list.listener || listener);
       } else if (typeof list !== 'function') {
         position = -1;

@ events.js:376 @ EventEmitter.prototype.removeAllListeners =
           this._events = new Map();
           this._eventsCount = 0;
         } else if (events.has(type)) {
-          if (--this._eventsCount === 0)
-            this._events = new Map();
-            // this._events.clear();
-          else
-            events.delete(type);
+          --this._eventsCount;
+          events.delete(type);
         }
         return this;
       }
```

Prof:
![image](https://user-images.githubusercontent.com/9109612/42839959-064f5244-8a0e-11e8-8865-9bb8a189fbf7.png)

prof json file (github doesn't like .json, so I renamed it to .txt)
[v8-clean-remove.txt](https://github.com/nodejs/node/files/2203015/v8-clean-remove.txt)